File: edu/cmu/hcii/puc/uigen/TwoColumnRow.java
/**
 * TwoColumnRow.java
 *
 * This node represents a row in the interface that uses a two column
 * layout, each with a single label and a single component.
 *
 * @author Jeffrey Nichols
 */

// Package Definition

package edu.cmu.hcii.puc.uigen;


// Import Declarations

import java.awt.Label;
import java.awt.Component;
import java.awt.Dimension;

import java.util.Vector;

import edu.cmu.hcii.puc.cio.ConcreteInteractionObject;
import edu.cmu.hcii.puc.cio.ContainerCIO;
import edu.cmu.hcii.puc.cio.LabelCIO;
import edu.cmu.hcii.puc.cio.StateLinkedCIO;


// Class Definition

public class TwoColumnRow extends Row {

    //**************************
    // Constants
    //**************************

    protected static final int LABEL1_INDEX = 0;
    protected static final int COMPONENT1_INDEX = 1;
    protected static final int LABEL2_INDEX = 2;
    protected static final int COMPONENT2_INDEX = 3;


    //**************************
    // Constructor
    //**************************

    public TwoColumnRow( PanelNode pPanel,
			 ConcreteInteractionObject pCIO1,
			 ConcreteInteractionObject pCIO2 ) {
	super( pPanel );

	addCIO( pCIO1.getLabelCIO() );
	addCIO( pCIO1 );

	addCIO( pCIO2.getLabelCIO() );
	addCIO( pCIO2 );
    }

    //**************************
    // Member Methods
    //**************************

    public void addComponents( ContainerCIO pContainer ) {
	LabelCIO pLabelCIO1 = (LabelCIO)m_vCIOs.elementAt( LABEL1_INDEX );
	StateLinkedCIO pStateCIO1 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT1_INDEX );
	LabelCIO pLabelCIO2 = (LabelCIO)m_vCIOs.elementAt( LABEL2_INDEX );
	StateLinkedCIO pStateCIO2 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT2_INDEX );

	pContainer.addCIO( pLabelCIO1 );
	pContainer.addCIO( pStateCIO1 );
	pContainer.addCIO( pLabelCIO2 );
	pContainer.addCIO( pStateCIO2 );
    }

    public Dimension getPreferredSize( LayoutVariables pVars ) {

	LabelCIO pLabelCIO1 = (LabelCIO)m_vCIOs.elementAt( LABEL1_INDEX );
	StateLinkedCIO pStateCIO1 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT1_INDEX );
	LabelCIO pLabelCIO2 = (LabelCIO)m_vCIOs.elementAt( LABEL2_INDEX );
	StateLinkedCIO pStateCIO2 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT2_INDEX );

	Dimension ret = new Dimension();

	ret.height = Math.max( pStateCIO1.getWidget().getPreferredSize().height,
			       pStateCIO2.getWidget().getPreferredSize().height ) + 
	             pVars.m_nRowPadding;
	ret.width = ( pLabelCIO1 != null ? 
		      pLabelCIO1.getWidget().getPreferredSize().width : 0 )
	            + pStateCIO1.getWidget().getPreferredSize().width + 
	            ( pLabelCIO2 != null ? 
	              pLabelCIO2.getWidget().getPreferredSize().width : 0 )
	            + pStateCIO2.getWidget().getPreferredSize().width ;


	return ret;
    }

    public int doLayout( ContainerCIO pContainer, 
			 int nTopY, 
			 LayoutVariables pVars ) {
	
	LabelCIO pLabelCIO1 = (LabelCIO)m_vCIOs.elementAt( LABEL1_INDEX );
	StateLinkedCIO pStateCIO1 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT1_INDEX );
	LabelCIO pLabelCIO2 = (LabelCIO)m_vCIOs.elementAt( LABEL2_INDEX );
	StateLinkedCIO pStateCIO2 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT2_INDEX );

	int nRowHeight = Math.max( pStateCIO1.getWidget().getPreferredSize().height,
				   pStateCIO2.getWidget().getPreferredSize().height );

	int nColumnWidth = ( pContainer.getWidget().getSize().width - 3 * pVars.m_nRowPadding ) / 2;
	int nCol1LabelWidth = (int)Math.round( nColumnWidth * pVars.m_dTwoColLabel1Pcnt );
	int nCol1CompWidth = nColumnWidth - nCol1LabelWidth;
	int nCol2LabelWidth = (int)Math.round( nColumnWidth * pVars.m_dTwoColLabel2Pcnt );
	int nCol2CompWidth = nColumnWidth - nCol2LabelWidth;

	Component pComp = pStateCIO1.getWidget();
	pComp.setSize( nCol1CompWidth,
		       nRowHeight );
	pComp.setLocation( nCol1LabelWidth + 2 * pVars.m_nRowPadding,
			   nTopY + pVars.m_nRowPadding );

	Label pLabel = null;
	if ( pLabelCIO1 != null ) {
	    pLabel = (Label)pLabelCIO1.getWidget();
	    pLabel.setAlignment( Label.RIGHT );
	    pLabel.setSize( nCol1LabelWidth,
			    nRowHeight );
	    pLabel.setLocation( pVars.m_nRowPadding, 
				nTopY + pVars.m_nRowPadding );
	}

	pComp = pStateCIO2.getWidget();
	pComp.setSize( nCol2CompWidth,
		       nRowHeight );
	pComp.setLocation( nColumnWidth + nCol2LabelWidth + 3 * pVars.m_nRowPadding,
			   nTopY + pVars.m_nRowPadding );

	pLabel = null;
	if ( pLabelCIO2 != null ) {
	    pLabel = (Label)pLabelCIO2.getWidget();
	    pLabel.setAlignment( Label.RIGHT );
	    pLabel.setSize( nCol2LabelWidth,
			    nRowHeight );
	    pLabel.setLocation( nColumnWidth + 2 * pVars.m_nRowPadding, 
				nTopY + pVars.m_nRowPadding );
	}

	return nTopY + nRowHeight + pVars.m_nRowPadding;
    }

    public String toString() {

	String ret = "TwoColumnRow\n";

	LabelCIO pLabelCIO1 = (LabelCIO)m_vCIOs.elementAt( LABEL1_INDEX );
	StateLinkedCIO pStateCIO1 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT1_INDEX );
	LabelCIO pLabelCIO2 = (LabelCIO)m_vCIOs.elementAt( LABEL2_INDEX );
	StateLinkedCIO pStateCIO2 = (StateLinkedCIO)m_vCIOs.elementAt( COMPONENT2_INDEX );

	if ( pLabelCIO1 != null ) 
	    ret += "0. LabelCIO1 - " + pLabelCIO1.getWidget().getBounds().toString() + "\n";
	ret += "1. StateCIO1 = " + pStateCIO1.getApplObj().m_sName + " - " + pStateCIO1.getWidget().getBounds().toString() + "\n";
	if ( pLabelCIO2 != null ) 
	    ret += "2. LabelCIO2 - " + pLabelCIO2.getWidget().getBounds().toString() + "\n";
	ret += "3. StateCIO2 = " + pStateCIO2.getApplObj().m_sName + " - " + pStateCIO2.getWidget().getBounds().toString() + "\n";


	return ret;
    }
}

File: csharp/UIGeneration/TwoColumnRow.cs
using System;
using System.Collections;
using System.Windows.Forms;
using PUC;
using PUC.CIO;

namespace PUC.UIGeneration
{
	/// <summary>
	/// Summary description for TwoColumnRow.
	/// </summary>
	public class TwoColumnRow : Row
	{
		/*
		 * Constants
		 */

		protected const int LABEL1_INDEX = 0;
		protected const int COMPONENT1_INDEX = 1;
		protected const int LABEL2_INDEX = 2;
		protected const int COMPONENT2_INDEX = 3;


		/*
		 * Constructor
		 */

		public TwoColumnRow( PanelNode panel, 
			ConcreteInteractionObject cio1, 
			ConcreteInteractionObject cio2 )
			: base( panel )
		{
			addCIO( cio1.GetLabelCIO() );
			addCIO( cio1 );
			addCIO( cio2.GetLabelCIO() );
			addCIO( cio2 );
		}


		/*
		 * Member Methods
		 */

		public override void AddComponents( ContainerCIO container )
		{
			LabelCIO labelCIO1 = (LabelCIO)_CIOs[ LABEL1_INDEX ];
			StateLinkedCIO stateCIO1 = (StateLinkedCIO)_CIOs[ COMPONENT1_INDEX ];
			LabelCIO labelCIO2 = (LabelCIO)_CIOs[ LABEL2_INDEX ];
			StateLinkedCIO stateCIO2 = (StateLinkedCIO)_CIOs[ COMPONENT2_INDEX ];

			container.AddCIO( labelCIO1 );
			container.AddCIO( stateCIO1 );
			container.AddCIO( labelCIO2 );
			container.AddCIO( stateCIO2 );
		}

		public override System.Drawing.Size GetPreferredSize(LayoutVariables vars)
		{
			LabelCIO labelCIO1 = (LabelCIO)_CIOs[ LABEL1_INDEX ];
			StateLinkedCIO stateCIO1 = (StateLinkedCIO)_CIOs[ COMPONENT1_INDEX ];
			LabelCIO labelCIO2 = (LabelCIO)_CIOs[ LABEL2_INDEX ];
			StateLinkedCIO stateCIO2 = (StateLinkedCIO)_CIOs[ COMPONENT2_INDEX ];

			System.Drawing.Size ret = new System.Drawing.Size( 0, 0 );

			ret.Height = Math.Max( stateCIO1.GetPreferredSize().Height,
								   stateCIO2.GetPreferredSize().Height );

			ret.Width = ( labelCIO1 != null ? labelCIO1.GetPreferredSize().Width : 0 ) +
						stateCIO1.GetPreferredSize().Width +
				        ( labelCIO2 != null ? labelCIO2.GetPreferredSize().Width : 0 ) +
						stateCIO2.GetPreferredSize().Width;

			ret.Height += vars.RowPadding;
			ret.Width += 3 * vars.RowPadding;

			return ret;			
		}

		public override int DoLayout(ContainerCIO container, int topY, LayoutVariables vars)
		{
			LabelCIO labelCIO1 = (LabelCIO)_CIOs[ LABEL1_INDEX ];
			StateLinkedCIO stateCIO1 = (StateLinkedCIO)_CIOs[ COMPONENT1_INDEX ];
			LabelCIO labelCIO2 = (LabelCIO)_CIOs[ LABEL2_INDEX ];
			StateLinkedCIO stateCIO2 = (StateLinkedCIO)_CIOs[ COMPONENT2_INDEX ];

			System.Windows.Forms.Label label = (Label)labelCIO1.GetControl();
			Control control = stateCIO1.GetControl();

			System.Drawing.Size prefSize1 = stateCIO1.GetPreferredSize();
			System.Drawing.Size prefSize2 = stateCIO2.GetPreferredSize();

			int rowHeight = Math.Max( prefSize1.Height, prefSize2.Height ); 
			int columnWidth = ( container.InternalSize.Width - 3 * vars.RowPadding ) / 2;
			int col1LabelWidth = (int)Math.Round( columnWidth * vars.TwoColLabel1Pcnt );
			int col1CompWidth = columnWidth - col1LabelWidth;
			int col2LabelWidth = (int)Math.Round( columnWidth * vars.TwoColLabel2Pcnt );
			int col2CompWidth = columnWidth = col2LabelWidth;

			control = stateCIO1.GetControl();
			control.Size = new System.Drawing.Size( col1CompWidth, rowHeight );
			control.Location = new System.Drawing.Point( col1LabelWidth + 2 * vars.RowPadding,
				topY - stateCIO1.GetControlOffset().Y );

			if ( labelCIO1 != null ) 
			{
				label = (Label)labelCIO1.GetControl();
				label.TextAlign = System.Drawing.ContentAlignment.TopRight;
				label.Size = new System.Drawing.Size( col1LabelWidth, rowHeight );
				label.Location = new System.Drawing.Point( vars.RowPadding,
					topY - labelCIO1.GetControlOffset().Y );
				labelCIO1.SetLabelText();
			}

			control = stateCIO2.GetControl();
			control.Size = new System.Drawing.Size( col2CompWidth, rowHeight );
			control.Location = new System.Drawing.Point( columnWidth + col2LabelWidth + 3 * vars.RowPadding,
				topY - stateCIO2.GetControlOffset().Y );

			if ( labelCIO2 != null ) 
			{
				label = (Label)labelCIO2.GetControl();
				label.TextAlign = System.Drawing.ContentAlignment.TopRight;
				label.Size = new System.Drawing.Size( col2LabelWidth, rowHeight );
				label.Location = new System.Drawing.Point( columnWidth + 2 * vars.RowPadding,
					topY - labelCIO2.GetControlOffset().Y );
				labelCIO2.SetLabelText();
			}

			return 1;
		}

		public override string ToString()
		{
			String ret = "TwoColumnRow\r\n";

			LabelCIO labelCIO1 = (LabelCIO)_CIOs[ LABEL1_INDEX ];
			StateLinkedCIO stateCIO1 = (StateLinkedCIO)_CIOs[ COMPONENT1_INDEX ];
			LabelCIO labelCIO2 = (LabelCIO)_CIOs[ LABEL2_INDEX ];
			StateLinkedCIO stateCIO2 = (StateLinkedCIO)_CIOs[ COMPONENT2_INDEX ];

			if ( labelCIO1 != null )
			{
				ret += "0. LabelCIO - [" + labelCIO1.GetControl().Bounds.X + "," + labelCIO1.GetControl().Bounds.Y + "," + labelCIO1.GetControl().Bounds.Width + "," + labelCIO1.GetControl().Bounds.Height + "]\r\n";
			}

			ret += "1. StateCIO1 = " + stateCIO1.GetApplObj().Name + " - [" + stateCIO1.GetControl().Bounds.X + "," + stateCIO1.GetControl().Bounds.Y + "," + stateCIO1.GetControl().Bounds.Width + "," + stateCIO1.GetControl().Bounds.Height + "]\r\n";

			ret += "2. LabelCIO - [" + labelCIO2.GetControl().Bounds.X + "," + labelCIO2.GetControl().Bounds.Y + "," + labelCIO2.GetControl().Bounds.Width + "," + labelCIO2.GetControl().Bounds.Height + "]\r\n";

			ret += "3. StateCIO2 = " + stateCIO2.GetApplObj().Name + " - [" + stateCIO2.GetControl().Bounds.X + "," + stateCIO2.GetControl().Bounds.Y + "," + stateCIO2.GetControl().Bounds.Width + "," + stateCIO2.GetControl().Bounds.Height + "]\r\n";

			return ret;
		}

	}
}
